### PR TITLE
Handle container run time version errors on a node more gracefully and faster.

### PR DESCRIPTION
**Pull request by aveshagarwal**
Changes in PR https://github.com/kubernetes/kubernetes/pull/19675 can cause version errors on downstream distros as they append distro string to docker version like 1.9.1-fc22. And following errors occur:

```
E0122 10:24:18.435492   27245 manager.go:986] docker: failed to parse docker server version "1.9.1-fc22": Unable to parse version "1.9.1-fc22": "1-fc22" is not an integer
E0122 10:24:23.273679   27245 manager.go:986] docker: failed to parse docker server version "1.9.1-fc22": Unable to parse version "1.9.1-fc22": "1-fc22" is not an integer
E0122 10:24:23.273692   27245 kubelet.go:2579] Container runtime sanity check failed: docker: failed to parse docker server version "1.9.1-fc22": Unable to parse version "1.9.1-fc22": "1-fc22" is not an integer
```

These error on a node should be handled more gracefully and faster as the current code does not take these errors into account. Because errors in the version cause container sanity check to fail, it takes almost 5 mins for the node to go down from Ready to NotReady state.

### CODE REVIEW

**Comment by aveshagarwal** [comment]
@ncdc @timstclair

**Comment by timothysc** [comment]
xref #20020

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -2770,16 +2770,19 @@ func (kl *Kubelet) setNodeStatusMachineInfo(node *api.Node) {
...
+		return err
 	} else {
 		node.Status.NodeInfo.KernelVersion = verinfo.KernelVersion
 		node.Status.NodeInfo.OSImage = verinfo.ContainerOsVersion
 
 		runtimeVersion := "Unknown"
-		if runtimeVer, err := kl.containerRuntime.Version(); err == nil {
+		if runtimeVer, err := kl.containerRuntime.Version(); err != nil {
+			return err
+		} else {
```
Comment (smarterclayton) [inline pkg/kubelet/kubelet.go]: Since you return here you can omit the else block (convention).

**Comment by smarterclayton** [comment]
@k8s-bot test this

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (aveshagarwal) [inline pkg/kubelet/kubelet.go]: But that would require declaring runtimeVer and err outside of if.

Comment (aveshagarwal) [inline pkg/kubelet/kubelet.go]: I have pushed an updated version but I feel that "if-else" would have been better as it does not increase the scope of runtimeVer and err beyond if-else as that is not needed.

**Comment by aveshagarwal** [comment]
Pushed updated PR.

**Comment by aveshagarwal** [comment]
```
Detected 0 ready nodes, found 0 nodes out of expected 6. Your cluster may not be fully functional.
2016/01/26 22:22:43 e2e.go:309: Error running up: exit status 2
2016/01/26 22:22:43 e2e.go:305: Step 'up' finished in 27m47.321809852s
2016/01/26 22:22:43 e2e.go:141: Error starting e2e cluster. Aborting.
exit status 1
```

**Comment by aveshagarwal** [comment]
Rebased.

**Comment by dchen1107** [comment]
There is cluster up issue, and it might be caused by your change here since the new logic is skip . Unfortunately the related logs are not collected. Can you double confirm with this?

**Comment by aveshagarwal** [comment]
I am not sure why this PR should cause the issue, because it only returns error when there is an issue with docker version string. Is there any issue with the docker version string?

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -2897,25 +2897,29 @@ func (kl *Kubelet) setNodeStatusMachineInfo(node *api.Node) {
 }
 
 // Set versioninfo for the node.
-func (kl *Kubelet) setNodeStatusVersionInfo(node *api.Node) {
+func (kl *Kubelet) setNodeStatusVersionInfo(node *api.Node) error {
 	verinfo, err := kl.cadvisor.VersionInfo()
 	if err != nil {
 		glog.Errorf("Error getting version info: %v", err)
+		return err
```
Comment (smarterclayton) [inline pkg/kubelet/kubelet.go]: style, remove the else (not needed anymore since you return early)

**Comment by aveshagarwal** [comment]
@smarterclayton thanks. Updated PR.

**Pull request closed without merging**